src/explore-education-statistics-admin/src/pages/release/edit-release/content/ReleaseContext.tsx

```diff
@@ -0,0 +1,237 @@
+import produce from 'immer';
+import {
+  EditableContentBlock,
+  ExtendedComment,
+  EditableRelease,
+} from '@admin/services/publicationService';
+import { DataBlock } from '@common/services/dataBlockService';
+import { ContentSection } from '@common/services/publicationService';
+import React, { createContext, useContext, useReducer } from 'react';
+import ReleaseDispatchAction from './actions';
+
+type Dispatch = (action: ReleaseDispatchAction) => void;
+type State = {
+  release: EditableRelease | undefined;
+  canUpdateRelease: boolean;
+  availableDataBlocks: DataBlock[];
+  unresolvedComments: ExtendedComment[];
+};
+type ReleaseProviderProps = { children: React.ReactNode };
+
+const ReleaseStateContext = createContext<State | undefined>(undefined);
+const ReleaseDispatchContext = createContext<Dispatch | undefined>(undefined);
+
+export function releaseReducer(state: State, action: ReleaseDispatchAction) {
+  switch (action.type) {
+    case 'CLEAR_STATE':
+      return {
+        release: undefined,
+        canUpdateRelease: false,
+        availableDataBlocks: [],
+        unresolvedComments: [],
+      };
+    case 'SET_STATE':
+    case 'SET_AVAILABLE_DATABLOCKS':
+      return { ...state, ...action.payload };
+    case 'REMOVE_BLOCK_FROM_SECTION':
+      return produce<State>(state, draft => {
+        const { sectionId, blockId, sectionKey } = action.payload.meta;
+        if (
+          draft.release === undefined ||
+          typeof draft.release[sectionKey] !== 'object'
+        ) {
+          throw new Error('REMOVE_BLOCK_FROM_SECTION: failed');
+        } else if (sectionKey === 'content') {
+          draft.release[sectionKey] = draft.release[sectionKey].map(section => {
```

Comment:
This isn't really idiomatic use of Immer. We can simplify this with something like (note I've used Lodash's `remove` function):

```javascript
import remove from 'lodash/remove';

const matchingSection = draft.release[sectionKey].find(section => section.id === sectionId);

if (matchingSection) {
  const blockIndex = matchingSection.content?.findIndex(contentBlock => contentBlock.id === blockId);

  remove(matchingSection, content => content.id === blockId);
}
```

The point of Immer is that we can mutate state freely within the producer function, so we can access a bunch of functions that would normally be a no-no for normal immutable code.